Settings: add new performance tab

keoshi opened the pull request:
#### Changes proposed in this Pull Request

* **This applies to both Jetpack and wpcom simple sites**
* Adds new Performance tab to settings.
* Moves the following modules to this new tab:
  * `Performance & Speed` (Jetpack only, previously on Writing)
  * `Media`: Video and Lazy-load (JP 5.8-alpha up) (Jetpack only, previously on Writing)
  * `Jetpack Search` (previously on Traffic)
  * `Accelerated Mobile Pages (AMP)` (previously on Traffic)

#### Notes

* Initial issue: https://github.com/Automattic/jetpack/issues/10550
* PR for Jetpack: https://github.com/Automattic/jetpack/pull/10941
* This will need updated e2e tests.

#### Before

![image](https://user-images.githubusercontent.com/390760/50222065-fad4cc00-038e-11e9-90dc-9fe6e2efced0.png)


#### After - Jetpack free

<!-- ![image](https://user-images.githubusercontent.com/390760/50222107-22c42f80-038f-11e9-8187-1ed6b2a6eb0f.png) -->

![image](https://user-images.githubusercontent.com/390760/50647446-7b3f2380-0f70-11e9-9720-f3955dd784ae.png)


#### After - Jetpack Professional / Atomic

<!-- ![image](https://user-images.githubusercontent.com/390760/50222230-8cdcd480-038f-11e9-8673-33c931e6eb4c.png) -->

![image](https://user-images.githubusercontent.com/390760/50643623-8b520580-0f66-11e9-8b67-c4b6159346f3.png)



#### After - wpcom simple site

![image](https://user-images.githubusercontent.com/390760/50222158-4c7d5680-038f-11e9-9d91-eeb556882180.png)

#### Testing instructions

* Fire up this branch.
* Take a look at the settings:
  * Ensure you see the new Performance tab.
  * Confirm changing settings saves them successfully.
  * Check the other settings tabs and make sure the modules are not duplicated.

Fixes #29545

a8ci18n commented on client/my-sites/site-settings/navigation.jsx:
```diff
@@ -28,6 +28,7 @@ export class SiteSettingsNavigation extends Component {
 		const { translate } = this.props;
 		return {
 			general: translate( 'General', { context: 'settings screen' } ),
+			performance: translate( 'Performance', { context: 'settings screen' } ),
```
Hi! I've found a possible matching string that has already been translated **20** times:
`translate( 'Performance' )` *ES Score: 22*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/performance-settings-tab)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

:information_source: This string already exists without a **context**. Only add a context if the meaning of the string is very specific.

tyxla commented on client/my-sites/site-settings/settings-performance/main.jsx:
```diff
@@ -0,0 +1,97 @@
+ * Internal dependencies
+ */
+import Main from 'components/main';
+import DocumentHead from 'components/data/document-head';
```
🔤 ❓ :trollface:

keoshi replied:
:D https://github.com/Automattic/wp-calypso/pull/29597/commits/bfdf02ba12ca859c9537394da6e42bbd0f1e7a5b

tyxla commented on client/my-sites/site-settings/settings-performance/controller.js:
```diff
@@ -0,0 +1,19 @@
+/**
+ * Internal dependencies
+ */
+import PerformanceMain from 'my-sites/site-settings/settings-performance/main';
```
We could shorten `'my-sites/site-settings/settings-performance/main'` to just `./main`.

keoshi replied:
✅https://github.com/Automattic/wp-calypso/pull/29597/commits/de2eaa7a61f653b1fadbe62c5b05a21e7a526f5c

tyxla commented on client/my-sites/site-settings/settings-performance/controller.js:
```diff
@@ -0,0 +1,19 @@
+ */
+import PerformanceMain from 'my-sites/site-settings/settings-performance/main';
+
+export default {
```
These days we try to avoid introducing `export default { func1, func2 }` for several reasons - in a nutshell to avoid unexpected issues when importing, and for build performance reasons.

Instead you could have just `export function performance() { ... }` or `export const performance = () => {}`

keoshi replied:
Would it be something like this?

```
export function performance( context, next ) {
	context.primary = React.createElement( SiteSettingsPerformance );
	next();
}
```

Or do those have to be under a `return`?

tyxla replied:
Yes, exactly - no need to be under a `return`. Here's an example just for reference: https://github.com/Automattic/wp-calypso/blob/a356b1e1ca3aba77360db20543d449da5c70a531/client/me/account-close/controller.js#L13

keoshi replied:
Thanks for that! Done in https://github.com/Automattic/wp-calypso/pull/29597/commits/608cd72f13180d9f7985db5e811ff6bb65149c31

tyxla commented on client/my-sites/site-settings/settings-performance/main.jsx:
```diff
@@ -0,0 +1,97 @@
+	return (
+		<Main className="settings-performance site-settings">
+			<DocumentHead title={ translate( 'Site Settings' ) } />
+			<JetpackDevModeNotice />
```
Is there any of the modules whose settings we display here that would be unavailable in Jetpack dev mode? If there's no such module, we might want to omit this.

keoshi replied:
Great point. @jeherve can you help confirming all of these are available in dev mode?

tyxla commented on client/my-sites/site-settings/settings-performance/main.jsx:
```diff
@@ -0,0 +1,97 @@
+} );
+
+const getFormSettings = partialRight( pick, [
+	'jetpack_search_enabled',
```
Seems like we forgot to remove those from `<SiteSettingsTraffic />`.

keoshi replied:
Damn, I could have sworn I did that. Anyway! https://github.com/Automattic/wp-calypso/pull/29597/commits/487a6c0826901faf2b940920cb2b3cdad582fe8a

tyxla commented on client/my-sites/site-settings/form-performance.jsx:
```diff
@@ -0,0 +1,131 @@
+export default flowRight(
+	connectComponent,
+	localize,
+	wrapSettingsForm( getFormSettings )
```
In `SiteSettingsPerformance` we already have a `wrapSettingsForm`. I'd suggest using it instead of introducing a nested form in here. What do you think? If you agree, it's worth to mention that all the settings we embed here can be moved directly to `<SiteSettingsPerformance />`, so `<SiteSettingsFormPerformance />` won't even be necessary.

keoshi commented on client/my-sites/site-settings/form-performance.jsx:
```diff
@@ -0,0 +1,131 @@
+import QueryJetpackModules from 'components/data/query-jetpack-modules';
+
+class SiteSettingsFormPerformance extends Component {
+	renderSectionHeader( title, showButton = true, disableButton = false ) {
```
True, I noticed the same, but found it easier to just duplicate for the sake of this PR. Would appreciate if you could do that later on!

tyxla replied:
Sure, I'd love to follow up on this. For now, as-is it's perfect for the purpose of this PR.

keoshi commented on client/my-sites/site-settings/navigation.jsx:
```diff
@@ -56,6 +57,14 @@ export class SiteSettingsNavigation extends Component {
 						{ strings.general }
 					</NavItem>
 
+					<NavItem
```
I also feel it's a little weird to have it before Writing _but_:

- It might be because I'm already used to it.
- Performance is arguably more important than the writing set up.

What do you think about this point @joanrho @jeffgolenski ?

joanrho replied:
Given customer feedback about how important site performance is to them (specifically with Jetpack), I'd think Performance should be the first tab immediately after General.

keoshi commented on client/my-sites/site-settings/form-performance.jsx:
```diff
@@ -0,0 +1,131 @@
+			<form
+				id="site-settings"
+				onSubmit={ handleSubmitForm }
+				className="site-settings__security-settings"
```
Yep, great catch. Fixed in https://github.com/Automattic/wp-calypso/pull/29597/commits/2d96f0d3dfe56ca90ce9fe067bad82086e9640b4

keoshi commented on client/my-sites/site-settings/form-performance.jsx:
```diff
@@ -0,0 +1,131 @@
+/**
+ * Internal dependencies
+ */
+import wrapSettingsForm from './wrap-settings-form';
```
Done in https://github.com/Automattic/wp-calypso/pull/29597/commits/1e3b71fab3d7cff5e43e4dc5db52bca56aa2a93a

keoshi commented on client/my-sites/site-settings/form-performance.jsx:
```diff
@@ -0,0 +1,131 @@
+				{ jetpackSettingsUI && (
+					<div>
+						{ this.renderSectionHeader( translate( 'Media' ) ) }
+						<MediaSettings
```
Good point. Should we split up “Media” into two different cards, @joanrho @jeffgolenski ?

joanrho replied:
Yes, that makes sense!

keoshi replied:
Let's address this in a separate PR.

tyxla replied:
Sounds good to me

jeffgolenski replied:
Let's make sure we're keeping it consistent with how its presented in jetpacks settings in 2p-admin. We've got a release going out soon with the new UI and that's concrete until the next version!

keoshi replied:
This was addressed in this PR — we should probably spin up a quick one on JP to make sure this change is replicated there.

keoshi commented on client/my-sites/site-settings/form-performance.jsx:
```diff
@@ -0,0 +1,131 @@
+		} = this.props;
+
+		return (
+			<form
```
I honestly don't know — can we get rid of it?

tyxla replied:
I think we totally can 😉

keoshi commented on client/my-sites/site-settings/form-performance.jsx:
```diff
@@ -0,0 +1,131 @@
+const connectComponent = connect( state => {
+	const siteId = getSelectedSiteId( state );
+	const siteIsJetpack = isJetpackSite( state, siteId );
+	const jetpackSettingsUiSupported = siteSupportsJetpackSettingsUi( state, siteId );
```
Awesome, thank you!

keoshi commented on client/my-sites/site-settings/form-performance.jsx:
```diff
@@ -0,0 +1,131 @@
+
+				{ jetpackSettingsUI &&
+					jetpackVersionSupportsLazyImages && (
+						<div>
```
Done in https://github.com/Automattic/wp-calypso/pull/29597/commits/c437b222db6832a213a3070a3466bebaf146ac48

keoshi commented on client/my-sites/site-settings/form-performance.jsx:
```diff
@@ -0,0 +1,131 @@
+import QueryJetpackModules from 'components/data/query-jetpack-modules';
+
+class SiteSettingsFormPerformance extends Component {
+	renderSectionHeader( title, showButton = true, disableButton = false ) {
```
Great catch; done in https://github.com/Automattic/wp-calypso/pull/29597/commits/123c887c5cdcd3e61996553bd168978bdf05883a

keoshi commented on client/my-sites/site-settings/form-performance.jsx:
```diff
@@ -0,0 +1,131 @@
+				onSubmit={ handleSubmitForm }
+				className="site-settings__security-settings"
+			>
+				<QueryJetpackModules siteId={ this.props.siteId } />
```
Done in https://github.com/Automattic/wp-calypso/pull/29597/commits/53c2e77107916ebf8408ed259b40197c5be9eb50

keoshi commented on client/my-sites/site-settings/form-performance.jsx:
```diff
@@ -0,0 +1,131 @@
+				onSubmit={ handleSubmitForm }
+				className="site-settings__security-settings"
+			>
+				<QueryJetpackModules siteId={ this.props.siteId } />
```
Done in https://github.com/Automattic/wp-calypso/pull/29597/commits/53c2e77107916ebf8408ed259b40197c5be9eb50

keoshi commented on client/my-sites/site-settings/form-performance.jsx:
```diff
@@ -0,0 +1,131 @@
+} );
+
+const getFormSettings = partialRight( pick, [
+	'photon',
```
Done in https://github.com/Automattic/wp-calypso/pull/29597/commits/2a8795d18dbd6c4708ad21bb236b3d2ed6053109

keoshi commented on client/my-sites/site-settings/settings-performance/controller.js:
```diff
@@ -0,0 +1,19 @@
+/**
+ * Internal dependencies
+ */
+import PerformanceMain from 'my-sites/site-settings/settings-performance/main';
```
Done in https://github.com/Automattic/wp-calypso/pull/29597/commits/183c46125e6287f246039e509b63e6e0eeee30a0

keoshi commented on client/my-sites/site-settings/form-performance.jsx:
```diff
@@ -0,0 +1,131 @@
+/** @format */
```
Cleaned it up in https://github.com/Automattic/wp-calypso/pull/29597/commits/e3d27153bb0274104c5c5a7ac13e413c91e429eb

keoshi commented on client/my-sites/site-settings/settings-performance/index.js:
```diff
@@ -0,0 +1,25 @@
+ */
+import controller from './controller';
+import { navigation, siteSelection } from 'my-sites/controller';
+import settingsController from 'my-sites/site-settings/settings-controller';
```
Same for `controller` => `{performance}`, right?

tyxla replied:
Yes, totally 👍 💯

keoshi replied:
Done in https://github.com/Automattic/wp-calypso/pull/29597/commits/fe888689b6162be3030402584ee7ca6970f96f63

keoshi commented on client/my-sites/site-settings/settings-performance/index.js:
```diff
@@ -0,0 +1,25 @@
+/**
+ * Internal dependencies
+ */
+import controller from './controller';
```
Done in https://github.com/Automattic/wp-calypso/pull/29597/commits/fe888689b6162be3030402584ee7ca6970f96f63

tyxla commented on client/my-sites/site-settings/settings-performance/index.js:
```diff
@@ -0,0 +1,25 @@
+import { makeLayout, render as clientRender } from 'controller';
+import { navigation, siteSelection } from 'my-sites/controller';
+import { performance } from './controller';
+import { siteSettings } from 'my-sites/site-settings/settings-controller';
```
I think I misleaded you here, assuming that the settings controller exports functions. But it doesn't: https://github.com/Automattic/wp-calypso/blob/7b91038e6b42d9fc63722cdc051d76e95e5ba698/client/my-sites/site-settings/settings-controller.js#L19

To fix this, we have to change this import back to:

```
import settingsController from 'my-sites/site-settings/settings-controller';
```

and refer to the controller as `settingsController.siteSettings` again.

keoshi commented:
@tyxla Just finished updating this PR with the latest changes:

### Summary

- Simplified the Performance main component where all relevant modules are now called directly.
- Carousel settings now live in Settings > Writing as suggested before.

### Details

- Got rid of `<SiteSettingsFormPerformance />` and all settings modules are now inside `<SiteSettingsPerformance />` [discussion](https://github.com/Automattic/wp-calypso/pull/29597#discussion_r243607372) — https://github.com/Automattic/wp-calypso/pull/29597/commits/f3fff9a9fd43e14cda7c57baa2e0edc41a95db39
- Split Media Settings into two different components: `Writing` and `Performance` [discussion](https://github.com/Automattic/wp-calypso/pull/29597#discussion_r243592958) — https://github.com/Automattic/wp-calypso/pull/29597/commits/d3797477826ce4771ba3aa42bd43da51dca5ada1
- Included the new `MediaSettingsWriting` component in Settings > Writing — https://github.com/Automattic/wp-calypso/pull/29597/commits/19993beebda3302517f8e1027049d87b97cdce35

keoshi commented:
This is now ready for review, @tyxla — Thank you once again!

keoshi commented on client/my-sites/site-settings/settings-performance/main.jsx:
```diff
@@ -0,0 +1,164 @@
+
+				{ jetpackSettingsUI && (
+					<Fragment>
+						{ this.renderSectionHeader( translate( 'Media' ) ) }
```
Thanks for noticing that — fixed in https://github.com/Automattic/wp-calypso/pull/29597/commits/c8cb13b719fad15e4ebe00af09c27b904beb68f6

keoshi commented on client/my-sites/site-settings/style.scss:
```diff
@@ -233,6 +233,7 @@
 
 .site-settings__analytics-settings,
 .site-settings__general-settings,
+.site-settings__performance-settings,
```
I believe this lived in the `SiteSettingsFormPerformance` that was deleted since then; I've added back the class on https://github.com/Automattic/wp-calypso/pull/29597/commits/d55d2188ce260326101690bc8dc94a1261b9f520

keoshi commented:
@tyxla Should be good to go!

tyxla commented on client/my-sites/site-settings/media-settings-writing.jsx:
```diff
@@ -0,0 +1,153 @@
+import { getSelectedSiteId } from 'state/ui/selectors';
+import { getSiteSlug } from 'state/sites/selectors';
+import QueryJetpackConnection from 'components/data/query-jetpack-connection';
+import classNames from 'classnames';
```
Nit: this is an external import and should be moved up to the "External dependencies" section.

tyxla commented on client/my-sites/site-settings/media-settings-writing.jsx:
```diff
@@ -0,0 +1,153 @@
+
+	render() {
+		const {
+			carouselActive,
```
Wanna 🔤 these too before 🚢 ?

tyxla commented on client/my-sites/site-settings/media-settings-writing.jsx:
```diff
@@ -0,0 +1,153 @@
+import FormLabel from 'components/forms/form-label';
+import CompactFormToggle from 'components/forms/form-toggle/compact';
+import SupportInfo from 'components/support-info';
+import isJetpackModuleActive from 'state/selectors/is-jetpack-module-active';
```
🔤 ? 😉

keoshi commented:
The latest commit addresses all three issues pointed out above: https://github.com/Automattic/wp-calypso/pull/29597/commits/c734530d44051f21a50b8dcdf2f8d3d80ffd57c2
